src/LiveryBridge.ts

```diff
@@ -0,0 +1,296 @@
+            },
+          });
+        });
+      })
+      .catch(() => Promise.reject(new Error('Handshake not complete.')));
+  }
+
+  public sendSubscribe<ValueType>(
+    name: string,
+    listener: (value: ValueType) => void,
+    arg?: string,
+  ) {
+    const id = uuid();
+    this.listenerMap.set(id, listener);
+    return this.sendCommand<string>(name, id, arg);
+  }
+
+  // eslint-disable-next-line class-methods-use-this, @typescript-eslint/no-unused-vars -- Not implemented here.
+  protected handleCommand(message: CommandMessage) {
+    // Should be implemented in child class.
+  }
+
+  protected handleHandshake(message: HandshakeMessage) {
+    if (message.version !== this.version) {
+      this.sendReject(message.id, new Error('Versions do not correspond.'));
+    } else {
+      const deferred = this.deferredMap.get(this.handshakeId);
+      if (deferred) {
+        deferred.reject(new Error('Received new handshake.'));
+      }
+      this.sendResolve(message.id, this.version);
+    }
+  }
+
+  protected handleMessage(message: LiveryMessage) {
+    if (this.messageListener) {
+      this.messageListener(
+        `Incoming Message:${JSON.stringify(message, null, 1)} \n\n`,
+      );
+    }
+    if (LiveryBridge.isHandshakeMessage(message)) {
+      this.handleHandshake(message);
+      return;
+    }
+    if (LiveryBridge.isResolveMessage(message)) {
+      const deferred = this.deferredMap.get(message.id);
+      if (deferred) {
+        deferred.resolve(message.value);
+        this.deferredMap.delete(message.id);
+      }
+      return;
+    }
+    if (LiveryBridge.isRejectMessage(message)) {
+      const deferred = this.deferredMap.get(message.id);
+      if (deferred) {
+        deferred.reject(message.error);
+        this.deferredMap.delete(message.id);
+      }
+      return;
+    }
+
+    if (LiveryBridge.isCommandMessage(message)) {
+      this.handleCommand(message);
+    }
+
+    if (LiveryBridge.isEventMessage(message)) {
+      const listener = this.listenerMap.get(message.id);
+      if (listener) {
+        listener(message.value);
+      }
+      return;
+    }
+
+    // should always be last
+    this.sendReject(message.id || 'unknown', new Error('Unknown Command'));
+  }
+
+  protected sendEvent<ValueType>(id: string, value: ValueType) {
+    return this.handshakePromise.then(() => {
+      this.sendMessage({
+        isLivery: true,
+        type: 'event',
+        value,
+        id,
+      });
+    });
+  }
+
+  protected sendHandshake(id: string, version: string) {
+    this.sendMessage({
+      isLivery: true,
+      type: 'handshake',
+      id,
+      version,
+    });
+
+    return new Promise<void>((resolve, reject) => {
+      this.deferredMap.set(id, {
+        resolve: () => {
+          resolve();
+        },
+        reject: (error: Error) => {
+          reject(error);
+        },
+      });
+    });
+  }
+
+  protected sendMessage<T extends LiveryMessage>(message: T) {
```

Comment:
I'd like to see this as: `sendMessage(type: string, properties: Record<string, unknown>)` or something like that and then just have it construct the message object: `{ isLivery: true, type, ...properties }`. Likewise I suppose we could just specify id as an optional argument and have this create the id for you if not specified and return that id to be used from where this is called? Reduce the isLivery repetition etc.